MAINT: (additional) array copy semantics shims

tylerjereddy opened the pull request:
* Fixes gh-20170

* restore the old `copy` behavior when using NumPy >= `2.0.0dev0` for now; I think Mateusz
was using `asarray` to solve the same problem more concisely previously--I'm not bothered if folks prefer that, this was just more obviously restoring the original behavior to me (at least, I think it is...)

* `test_raise_non_numeric_gh18254` was mocking an array-like object so it gained an otherwise
unused `copy` argument for compliance; same goes
for `FakeArray2` in FFT testing (and in other
places that use arr mocking after that...)

* old shims using `np._CopyMode` are not safe to use (well, they appear to be more complex
than just bools now)

* some non-mock objects (I think) like `LowRankMatrix` and `Jacobian` were also given a `copy` argument for arr to play along, though I haven't thought about the deeper consequences of doing that

* `array_api_strict` needs a `copy` shim or two as well I think; that's external and not in scope here I don't think, but will cause this to fail until then:
`test_dispatch_to_unrecognize_library`

* apart from that one test vs. NumPy `main`, the rest of the suite passed with NumPy `main` and `1.26.4` on this branch in my hands

[skip circle] [skip cirrus]

@mtsokol I'm happy to back out the stuff you already covered in gh-20171, I think that's just a tiny subset of this though, so shouldn't be too disruptive.

tylerjereddy committed:
MAINT: array copy semantics shims

* Fixes gh-20170

* restore the old `copy` behavior when using
NumPy >= `2.0.0dev0` for now; I think Mateusz
was using `asarray` to solve the same problem more concisely
previously--I'm not bothered if folks prefer that,
this was just more obviously restoring the original
behavior to me (at least, I think it is...)

* `test_raise_non_numeric_gh18254` was mocking
an array-like object so it gained an otherwise
unused `copy` argument for compliance; same goes
for `FakeArray2` in FFT testing (and in other
places that use arr mocking after that...)

* old shims using `np._CopyMode` are not safe
to use (well, they appear to be more complex
than just bools now)

* some non-mock objects (I think) like `LowRankMatrix` and `Jacobian`
were also given a `copy` argument for arr to play along,
though I haven't thought about the deeper consequences
of doing that

* `array_api_strict` needs a `copy` shim or two as well
I think; that's external and not in scope here I don't think,
but will cause this to fail until then:
`test_dispatch_to_unrecognize_library`

* apart from that one test vs. NumPy `main`, the rest of
the suite passed with NumPy `main` and `1.26.4` on this
branch in my hands

[skip circle] [skip cirrus]

mtsokol commented:
> @mtsokol I'm happy to back out the stuff you already covered in https://github.com/scipy/scipy/pull/20171, I think that's just a tiny subset of this though, so shouldn't be too disruptive.

@tylerjereddy Let's keep it all here in one PR! I see that `__array__(..., copy=...)` change was missing - this is something new that was added to NumPy `copy` PR after I did SciPy changes - I forgot to address those too.

rgommers committed:
MAINT: add `copy_if_needed` to `scipy._lib`

Also make the code to deal with changes in NumPy around the `copy`
keyword as concise as possible; do the version checks only in a single
place.

This also resolved linting errors in the previous commit.

[skip cirrus]

rgommers committed:
MAINT: also add `dtype` parameter to `__array__` to complete signatures

This is required, not doing so is deprecated. Unclear why it doesn't
always emit a warning.

Also reduce overparametrization of a stats test that used `__array__`.

[skip cirrus]

rgommers commented:
Thanks @tylerjereddy. I added some changes to:
- resolve linter complaints,
- complete signatures of `__array__` so they always have both `dtype` and `copy` keywords
- do the checks for numpy version only once and make the inline checks for `copy` keywords as concise as possible
- removed some over-parametrization of a `stats` test that used `__array__`, since I noticed it and it failed ~20x with the same issue

lucascolley commented:
Remaining CI failure is a real complaint from Mypy

rgommers committed:
TYP: fix Mypy complaint due to try/except that is numpy-version-dependent

[skip cirrus] [skip circle]

rgommers commented:
> Remaining CI failure is a real complaint from Mypy

Fixed now.

This is probably best squash-merged once CI is happy. I'm away for most of the day, but if whoever does the merge could also hit the "Run workflow" button on https://github.com/scipy/scipy/actions/workflows/wheels.yml after merging, that would be very useful.

lucascolley commented on scipy/optimize/_nonlin.py:
```diff
@@ -675,7 +679,7 @@ def append(self, c, d):
             self.collapse()
 
-    def __array__(self):
+    def __array__(self, copy=None):
```
`dtype=None` is used in every other signature change in the PR but not here - is that deliberate? Otherwise LGTM

rgommers replied:
Good catch. Tests didn't fail and my code search skills apparently fell short.

This code doesn't really seem to be used, I'll add the keyword with a warning that any input that's not `None` is not handled.

rgommers committed:
MAINT: fix one more location which was missing dtype=None

[ci skip]

rgommers commented:
We have several broken CI jobs and wheels to this, so I hit the green button. Restarting wheel builds now. Thanks @tylerjereddy and @lucascolley!
